bundles/auth/org.eclipse.smarthome.auth.oauth2client/src/main/java/org/eclipse/smarthome/auth/oauth2client/internal/OAuthStoreHandlerImpl.java

```diff
@@ -0,0 +1,540 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.auth.oauth2client.internal;
+
+import static org.eclipse.smarthome.auth.oauth2client.internal.StorageRecordType.*;
+
+import java.security.GeneralSecurityException;
+import java.time.LocalDateTime;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.eclipse.jdt.annotation.NonNull;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.auth.oauth2client.internal.OAuthClientServiceImpl.PersistedParams;
+import org.eclipse.smarthome.auth.oauth2client.internal.cipher.SymmetricKeyCipher;
+import org.eclipse.smarthome.core.auth.client.oauth2.AccessTokenResponse;
+import org.eclipse.smarthome.core.auth.client.oauth2.StorageCipher;
+import org.eclipse.smarthome.core.storage.Storage;
+import org.eclipse.smarthome.core.storage.StorageService;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.RemovalListener;
+import com.google.common.cache.RemovalNotification;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonDeserializer;
+import com.google.gson.JsonPrimitive;
+import com.google.gson.JsonSerializer;
+
+/**
+ * This class handles the storage directly. It is internal to the OAuthClientService and there is
+ * little need to study this.
+ *
+ * The first role of this handler storing and caching the access token response, and persisted parameters.
+ *
+ * The storage contains these:
+ * 1. INDEX_HANDLES = json string-set of all handles
+ * 2. <handle>.LastUsed = system-time-milliseconds
+ * 3. <handle>.AccessTokenResponse = Json of AccessTokenResponse
+ * 4. <handle>.ServiceConfiguration = Json of PersistedParameters
+ *
+ * If at any time, the storage is not available, it is still possible to read existing access tokens from store.
+ * The last-used statistics for this access token is broken. It is a measured risk to take.
+ *
+ * If at any time, the storage is not available, it is not able to write any new access tokens into store.
+ *
+ * All entries are subject to removal if they have not been used for 183 days or more (half year).
+ * The recycle is performed when then instance is deactivated
+ *
+ * @author Gary Tse - Initial Contribution
+ *
+ */
+@Component(service = OAuthStoreHandler.class, property = {
```

Comment:
you can remove service attribute since it is automatically added and by default it will use all the implemented interfaces.

Reply:
Fixed